File: win_assist_ai.js
#!/usr/bin/env node

/**
 * WinAssistAI - Cross-Platform Node.js Launcher
 * Universal launcher that works on any system with Node.js
 *
 * Usage:
 *   node win_assist_ai.js [options]
 *   npm start [-- options]
 */

const { spawn, exec } = require('child_process');
const path = require('path');
const fs = require('fs');
const os = require('os');

// Colors for console output
const colors = {
    reset: '\x1b[0m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m',
    white: '\x1b[37m'
};

class WinAssistAILauncher {
    constructor() {
        this.scriptDir = __dirname;
        this.platform = this.detectPlatform();
        this.powershellCmd = null;
    }

    // Colored output functions
    log(color, message) {
        console.log(`${colors[color]}${message}${colors.reset}`);
    }

    error(message) { this.log('red', `[ERROR] ${message}`); }
    success(message) { this.log('green', `[✓] ${message}`); }
    warning(message) { this.log('yellow', `[WARNING] ${message}`); }
    info(message) { this.log('cyan', `[INFO] ${message}`); }
    debug(message) { this.log('blue', `[DEBUG] ${message}`); }

    // Display banner
    showBanner() {
        this.log('cyan', `
 ██╗    ██╗██╗███╗   ██╗ █████╗ ███████╗███████╗██╗███████╗████████╗ █████╗ ██╗
 ██║    ██║██║████╗  ██║██╔══██╗██╔════╝██╔════╝██║██╔════╝╚══██╔══╝██╔══██╗██║
 ██║ █╗ ██║██║██╔██╗ ██║███████║███████╗███████╗██║███████╗   ██║   ███████║██║
 ██║███╗██║██║██║╚██╗██║██╔══██║╚════██║╚════██║██║╚════██║   ██║   ██╔══██║██║
 ╚███╔███╔╝██║██║ ╚████║██║  ██║███████║███████║██║███████║   ██║   ██║  ██║██║
  ╚══╝╚══╝ ╚═╝╚═╝  ╚═══╝╚═╝  ╚═╝╚══════╝╚══════╝╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝

                    Voice-Controlled Windows Assistant
                    Node.js Cross-Platform Universal Launcher
        `);
    }

    // Detect platform and environment
    detectPlatform() {
        const platform = os.platform();
        const env = process.env;
        
        if (platform === 'win32') {
            if (env.MSYSTEM) {
                return { type: 'git-bash', name: `Git Bash (${env.MSYSTEM})` };
            } else if (env.WSL_DISTRO_NAME) {
                return { type: 'wsl', name: `WSL (${env.WSL_DISTRO_NAME})` };
            } else {
                return { type: 'windows', name: 'Windows' };
            }
        } else if (platform === 'linux') {
            if (env.WSL_DISTRO_NAME) {
                return { type: 'wsl', name: `WSL (${env.WSL_DISTRO_NAME})` };
            } else {
                return { type: 'linux', name: 'Linux' };
            }
        } else if (platform === 'darwin') {
            return { type: 'macos', name: 'macOS' };
        } else {
            return { type: 'unknown', name: platform };
        }
    }

    // Check for PowerShell availability
    async checkPowerShell() {
        const candidates = ['powershell.exe', 'pwsh', 'powershell'];
        
        for (const cmd of candidates) {
            try {
                await this.executeCommand(`${cmd} -Version`, { timeout: 5000 });
                this.powershellCmd = cmd;
                return cmd;
            } catch (error) {
                // Try next candidate
            }
        }
        return null;
    }

    // Execute a command and return a promise
    executeCommand(command, options = {}) {
        return new Promise((resolve, reject) => {
            const timeout = options.timeout || 30000;
            const cwd = options.cwd || this.scriptDir;
            
            const child = exec(command, { 
                cwd, 
                timeout,
                windowsHide: true 
            }, (error, stdout, stderr) => {
                if (error) {
                    reject({ error, stdout, stderr });
                } else {
                    resolve({ stdout, stderr });
                }
            });

            // Handle timeout
            setTimeout(() => {
                child.kill();
                reject(new Error('Command timeout'));
            }, timeout);
        });
    }

    // Launch PowerShell script
    async launchPowerShellScript(scriptName, args = []) {
        if (!this.powershellCmd) {
            throw new Error('PowerShell not available');
        }

        const scriptPath = path.join(this.scriptDir, 'scripts', scriptName);
        
        // Convert path for Windows if needed
        let winScriptPath = scriptPath;
        if (this.platform.type === 'git-bash' && path.sep === '/') {
            winScriptPath = scriptPath.replace(/\//g, '\\');
        }

        const psArgs = [
            '-ExecutionPolicy', 'Bypass',
            '-File', `"${winScriptPath}"`
        ];

        if (args.length > 0) {
            psArgs.push(...args);
        }

        this.info(`Launching ${scriptName} with ${this.powershellCmd}...`);
        this.debug(`Command: ${this.powershellCmd} ${psArgs.join(' ')}`);

        return new Promise((resolve, reject) => {
            const child = spawn(this.powershellCmd, psArgs, {
                stdio: 'inherit',
                cwd: this.scriptDir,
                windowsHide: false
            });

            child.on('close', (code) => {
                if (code === 0) {
                    resolve(code);
                } else {
                    reject(new Error(`Script exited with code ${code}`));
                }
            });

            child.on('error', (error) => {
                reject(error);
            });
        });
    }

    // Show help
    showHelp() {
        this.showBanner();
        this.log('yellow', '\nUSAGE:');
        console.log('  node win_assist_ai.js [options]');
        console.log('  npm start [-- options]');
        console.log('');
        this.log('yellow', 'OPTIONS:');
        console.log('  -h, --help           Show this help message');
        console.log('  -v, --version        Show version information');
        console.log('  -t, --test-voice     Test text-to-speech functionality');
        console.log('  --test-elevenlabs    Test ElevenLabs AI voice system');
        console.log('  --list-voices        List available ElevenLabs voices');
        console.log('  --setup-elevenlabs   Setup ElevenLabs AI voice synthesis');
        console.log('  -l, --list-commands  List all available commands');
        console.log('  -s, --with-serenade  Force launch with Serenade integration');
        console.log('  -n, --no-serenade    Launch without Serenade integration');
        console.log('  --start-full         Launch full system with Serenade (recommended)');
        console.log('  --install-serenade   Install Serenade voice control');
        console.log('');
        this.log('cyan', 'PLATFORM COMPATIBILITY:');
        console.log('  ✓ Windows (native)');
        console.log('  ✓ Git Bash (MINGW64/MINGW32)');
        console.log('  ✓ Windows Command Prompt');
        console.log('  ✓ Windows PowerShell');
        console.log('  ✓ WSL (Windows Subsystem for Linux)');
        console.log('  ✓ Linux terminals (with PowerShell Core)');
        console.log('  ✓ macOS terminals (with PowerShell Core)');
        console.log('');
        this.log('magenta', 'ELEVENLABS AI VOICE:');
        console.log('  ✓ High-quality AI voice synthesis');
        console.log('  ✓ Natural-sounding speech generation');
        console.log('  ✓ Multiple voice options and personalities');
        console.log('  ✓ Automatic fallback to Windows TTS');
        console.log('');
        this.log('yellow', 'EXAMPLES:');
        console.log('  node win_assist_ai.js                    # Start with auto-detection');
        console.log('  node win_assist_ai.js --start-full       # Full startup with Serenade');
        console.log('  node win_assist_ai.js --test-voice       # Test voice functionality');
        console.log('  node win_assist_ai.js --setup-elevenlabs # Setup AI voice');
        console.log('  node win_assist_ai.js --test-elevenlabs  # Test AI voice system');
        console.log('  npm start                                # Start via npm');
        console.log('  npm start -- --help                     # Show help via npm');
        console.log('');
    }

    // Main execution function
    async run(args) {
        try {
            // Parse arguments
            const options = this.parseArgs(args);

            if (options.help) {
                this.showHelp();
                return;
            }

            if (options.version) {
                this.showBanner();
                console.log('WinAssistAI Node.js Launcher v1.0.0');
                console.log('Node.js Version:', process.version);
                console.log('Platform:', this.platform.name);
                return;
            }

            // Display banner and platform info
            this.showBanner();
            this.success(`Welcome to WinAssistAI!`);
            this.info(`Platform: ${this.platform.name}`);
            this.info(`Node.js: ${process.version}`);

            // Check PowerShell availability
            this.info('Checking PowerShell availability...');
            const psCmd = await this.checkPowerShell();
            
            if (!psCmd) {
                this.error('PowerShell not found!');
                this.warning('Please install PowerShell:');
                console.log('  • Windows: PowerShell is built-in');
                console.log('  • Linux: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux');
                console.log('  • macOS: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-macos');
                process.exit(1);
            }

            this.success(`PowerShell found: ${psCmd}`);

            // Execute based on options
            if (options.testVoice) {
                await this.launchPowerShellScript('win_assist_ai.ps1', ['-TestVoice']);
            } else if (options.testElevenlabs) {
                this.info('Testing ElevenLabs AI voice system...');
                await this.launchPowerShellScript('test-elevenlabs.ps1', ['-Test']);
            } else if (options.listVoices) {
                this.info('Listing available ElevenLabs voices...');
                await this.launchPowerShellScript('test-elevenlabs.ps1', ['-ListVoices']);
            } else if (options.setupElevenlabs) {
                this.info('Setting up ElevenLabs AI voice synthesis...');
                await this.launchPowerShellScript('setup-elevenlabs.ps1');
            } else if (options.listCommands) {
                await this.launchPowerShellScript('win_assist_ai.ps1', ['-ListCommands']);
            } else if (options.installSerenade) {
                this.info('Installing Serenade voice control...');
                await this.launchPowerShellScript('install-serenade.ps1');
            } else if (options.startFull) {
                this.info('Starting full system with Serenade integration...');
                await this.launchPowerShellScript('start-with-serenade.ps1');
            } else if (options.withSerenade) {
                this.info('Starting with Serenade integration...');
                await this.launchPowerShellScript('win_assist_ai.ps1', ['-WithSerenade']);
            } else if (options.noSerenade) {
                this.info('Starting without Serenade integration...');
                await this.launchPowerShellScript('win_assist_ai.ps1', ['-NoSerenade']);
            } else {
                this.info('Starting with auto-detection...');
                await this.launchPowerShellScript('win_assist_ai.ps1');
            }

            this.success('WinAssistAI launched successfully!');
            this.log('yellow', 'Use "node win_assist_ai.js --help" for more options');

        } catch (error) {
            this.error(`Failed to launch WinAssistAI: ${error.message}`);
            this.debug(`Error details: ${error.stack}`);
            process.exit(1);
        }
    }

    // Parse command line arguments
    parseArgs(args) {
        const options = {};
        
        for (let i = 0; i < args.length; i++) {
            const arg = args[i];
            
            switch (arg) {
                case '-h':
                case '--help':
                    options.help = true;
                    break;
                case '-v':
                case '--version':
                    options.version = true;
                    break;
                case '-t':
                case '--test-voice':
                    options.testVoice = true;
                    break;
                case '-l':
                case '--list-commands':
                    options.listCommands = true;
                    break;
                case '-s':
                case '--with-serenade':
                    options.withSerenade = true;
                    break;
                case '-n':
                case '--no-serenade':
                    options.noSerenade = true;
                    break;
                case '--start-full':
                    options.startFull = true;
                    break;
                case '--install-serenade':
                    options.installSerenade = true;
                    break;
                case '--test-elevenlabs':
                    options.testElevenlabs = true;
                    break;
                case '--list-voices':
                    options.listVoices = true;
                    break;
                case '--setup-elevenlabs':
                    options.setupElevenlabs = true;
                    break;
                default:
                    if (arg.startsWith('-')) {
                        console.warn(`Unknown option: ${arg}`);
                    }
            }
        }
        
        return options;
    }
}

// Main execution
if (require.main === module) {
    const launcher = new WinAssistAILauncher();
    const args = process.argv.slice(2);
    launcher.run(args);
}

module.exports = WinAssistAILauncher;

File: win_assist_ai.py
#!/usr/bin/env python3
"""
WinAssistAI - Universal Python Launcher
Cross-platform launcher that works on any system with Python 3.6+

Usage:
    python win_assist_ai.py [options]
    python3 win_assist_ai.py [options]
"""

import sys
import os
import platform
import subprocess
import argparse
import shutil
from pathlib import Path

class Colors:
    """ANSI color codes for terminal output"""
    RESET = '\033[0m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'

class WinAssistAILauncher:
    def __init__(self):
        self.script_dir = Path(__file__).parent.absolute()
        self.platform_info = self.detect_platform()
        self.powershell_cmd = None
        
    def log(self, color, message):
        """Print colored message to console"""
        print(f"{color}{message}{Colors.RESET}")
        
    def error(self, message):
        self.log(Colors.RED, f"[ERROR] {message}")
        
    def success(self, message):
        self.log(Colors.GREEN, f"[✓] {message}")
        
    def warning(self, message):
        self.log(Colors.YELLOW, f"[WARNING] {message}")
        
    def info(self, message):
        self.log(Colors.CYAN, f"[INFO] {message}")
        
    def debug(self, message):
        self.log(Colors.BLUE, f"[DEBUG] {message}")

    def show_banner(self):
        """Display the WinAssistAI banner"""
        banner = """
 ██╗    ██╗██╗███╗   ██╗ █████╗ ███████╗███████╗██╗███████╗████████╗ █████╗ ██╗
 ██║    ██║██║████╗  ██║██╔══██╗██╔════╝██╔════╝██║██╔════╝╚══██╔══╝██╔══██╗██║
 ██║ █╗ ██║██║██╔██╗ ██║███████║███████╗███████╗██║███████╗   ██║   ███████║██║
 ██║███╗██║██║██║╚██╗██║██╔══██║╚════██║╚════██║██║╚════██║   ██║   ██╔══██║██║
 ╚███╔███╔╝██║██║ ╚████║██║  ██║███████║███████║██║███████║   ██║   ██║  ██║██║
  ╚══╝╚══╝ ╚═╝╚═╝  ╚═══╝╚═╝  ╚═╝╚══════╝╚══════╝╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝

                    Voice-Controlled Windows Assistant
                    Python Universal Cross-Platform Launcher
        """
        self.log(Colors.CYAN, banner)

    def detect_platform(self):
        """Detect the current platform and environment"""
        system = platform.system().lower()
        env = os.environ
        
        if system == 'windows':
            if 'MSYSTEM' in env:
                return {
                    'type': 'git-bash',
                    'name': f"Git Bash ({env['MSYSTEM']})",
                    'system': 'windows'
                }
            elif 'WSL_DISTRO_NAME' in env:
                return {
                    'type': 'wsl',
                    'name': f"WSL ({env['WSL_DISTRO_NAME']})",
                    'system': 'linux'
                }
            else:
                return {
                    'type': 'windows',
                    'name': 'Windows',
                    'system': 'windows'
                }
        elif system == 'linux':
            if 'WSL_DISTRO_NAME' in env:
                return {
                    'type': 'wsl',
                    'name': f"WSL ({env['WSL_DISTRO_NAME']})",
                    'system': 'linux'
                }
            else:
                return {
                    'type': 'linux',
                    'name': 'Linux',
                    'system': 'linux'
                }
        elif system == 'darwin':
            return {
                'type': 'macos',
                'name': 'macOS',
                'system': 'darwin'
            }
        else:
            return {
                'type': 'unknown',
                'name': system,
                'system': system
            }

    def check_powershell(self):
        """Check for PowerShell availability"""
        candidates = ['powershell.exe', 'pwsh', 'powershell']
        
        for cmd in candidates:
            if shutil.which(cmd):
                try:
                    # Test if the command works
                    result = subprocess.run(
                        [cmd, '-Command', 'Write-Host "Test"'],
                        capture_output=True,
                        timeout=5,
                        text=True
                    )
                    if result.returncode == 0:
                        self.powershell_cmd = cmd
                        return cmd
                except (subprocess.TimeoutExpired, subprocess.SubprocessError):
                    continue
        return None

    def launch_powershell_script(self, script_name, args=None):
        """Launch a PowerShell script with the given arguments"""
        if not self.powershell_cmd:
            raise RuntimeError("PowerShell not available")
            
        script_path = self.script_dir / 'scripts' / script_name
        
        if not script_path.exists():
            raise FileNotFoundError(f"Script not found: {script_path}")
        
        # Build PowerShell command
        ps_args = [
            self.powershell_cmd,
            '-ExecutionPolicy', 'Bypass',
            '-File', str(script_path)
        ]
        
        if args:
            ps_args.extend(args)
        
        self.info(f"Launching {script_name} with {self.powershell_cmd}...")
        self.debug(f"Command: {' '.join(ps_args)}")
        
        try:
            # Run the PowerShell script
            result = subprocess.run(
                ps_args,
                cwd=str(self.script_dir),
                timeout=300  # 5 minute timeout
            )
            return result.returncode
        except subprocess.TimeoutExpired:
            self.error("Script execution timed out")
            return 1
        except KeyboardInterrupt:
            self.warning("Script execution interrupted by user")
            return 130

    def show_help(self):
        """Display help information"""
        self.show_banner()
        
        help_text = """
USAGE:
  python win_assist_ai.py [options]
  python3 win_assist_ai.py [options]

OPTIONS:
  -h, --help           Show this help message
  -v, --version        Show version information
  -t, --test-voice     Test text-to-speech functionality
  -l, --list-commands  List all available commands
  -s, --with-serenade  Force launch with Serenade integration
  -n, --no-serenade    Launch without Serenade integration
  --start-full         Launch full system with Serenade (recommended)
  --install-serenade   Install Serenade voice control

PLATFORM COMPATIBILITY:
  ✓ Windows (all versions)
  ✓ Git Bash (MINGW64/MINGW32)
  ✓ Windows Command Prompt
  ✓ Windows PowerShell
  ✓ WSL (Windows Subsystem for Linux)
  ✓ Linux (with PowerShell Core)
  ✓ macOS (with PowerShell Core)
  ✓ Any system with Python 3.6+

EXAMPLES:
  python win_assist_ai.py                    # Start with auto-detection
  python win_assist_ai.py --start-full       # Full startup with Serenade
  python win_assist_ai.py --test-voice       # Test voice functionality
  python win_assist_ai.py --list-commands    # Show all commands
  python win_assist_ai.py --install-serenade # Install Serenade

REQUIREMENTS:
  • Python 3.6 or later
  • PowerShell (built-in on Windows, installable on Linux/macOS)
  • Internet connection (for some features)
        """
        
        print(help_text)

    def run(self, args):
        """Main execution function"""
        try:
            # Parse arguments
            parser = argparse.ArgumentParser(
                description='WinAssistAI Universal Python Launcher',
                add_help=False  # We'll handle help ourselves
            )
            
            parser.add_argument('-h', '--help', action='store_true')
            parser.add_argument('-v', '--version', action='store_true')
            parser.add_argument('-t', '--test-voice', action='store_true')
            parser.add_argument('-l', '--list-commands', action='store_true')
            parser.add_argument('-s', '--with-serenade', action='store_true')
            parser.add_argument('-n', '--no-serenade', action='store_true')
            parser.add_argument('--start-full', action='store_true')
            parser.add_argument('--install-serenade', action='store_true')
            
            options = parser.parse_args(args)
            
            if options.help:
                self.show_help()
                return 0
                
            if options.version:
                self.show_banner()
                print("WinAssistAI Python Launcher v1.0.0")
                print(f"Python Version: {sys.version}")
                print(f"Platform: {self.platform_info['name']}")
                return 0
            
            # Display banner and platform info
            self.show_banner()
            self.success("Welcome to WinAssistAI!")
            self.info(f"Platform: {self.platform_info['name']}")
            self.info(f"Python: {sys.version.split()[0]}")
            
            # Check PowerShell availability
            self.info("Checking PowerShell availability...")
            ps_cmd = self.check_powershell()
            
            if not ps_cmd:
                self.error("PowerShell not found!")
                self.warning("Please install PowerShell:")
                print("  • Windows: PowerShell is built-in")
                print("  • Linux: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux")
                print("  • macOS: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-macos")
                return 1
            
            self.success(f"PowerShell found: {ps_cmd}")
            
            # Execute based on options
            exit_code = 0
            
            if options.test_voice:
                exit_code = self.launch_powershell_script('win_assist_ai.ps1', ['-TestVoice'])
            elif options.list_commands:
                exit_code = self.launch_powershell_script('win_assist_ai.ps1', ['-ListCommands'])
            elif options.install_serenade:
                self.info('Installing Serenade voice control...')
                exit_code = self.launch_powershell_script('install-serenade.ps1')
            elif options.start_full:
                self.info('Starting full system with Serenade integration...')
                exit_code = self.launch_powershell_script('start-with-serenade.ps1')
            elif options.with_serenade:
                self.info('Starting with Serenade integration...')
                exit_code = self.launch_powershell_script('win_assist_ai.ps1', ['-WithSerenade'])
            elif options.no_serenade:
                self.info('Starting without Serenade integration...')
                exit_code = self.launch_powershell_script('win_assist_ai.ps1', ['-NoSerenade'])
            else:
                self.info('Starting with auto-detection...')
                exit_code = self.launch_powershell_script('win_assist_ai.ps1')
            
            if exit_code == 0:
                self.success('WinAssistAI launched successfully!')
                self.log(Colors.YELLOW, 'Use "python win_assist_ai.py --help" for more options')
            else:
                self.error(f'WinAssistAI exited with code {exit_code}')
            
            return exit_code
            
        except KeyboardInterrupt:
            self.warning('Operation cancelled by user')
            return 130
        except Exception as e:
            self.error(f'Failed to launch WinAssistAI: {e}')
            return 1

def main():
    """Entry point for the script"""
    launcher = WinAssistAILauncher()
    sys.exit(launcher.run(sys.argv[1:]))

if __name__ == '__main__':
    main()